linkerd/proxy/http/src/timeout.rs

```diff
@@ -34,28 +29,35 @@ pub struct MakeFuture<F> {
     timeout: Option<Duration>,
 }
 
-#[derive(Clone, Debug)]
-pub struct Service<S>(Timeout<S>);
+impl<M> tower::layer::Layer<M> for MakeTimeoutLayer {
+    type Service = MakeTimeout<M>;
 
-/// A marker set in `http::Response::extensions` that *this* process triggered
-/// the request timeout.
-#[derive(Debug)]
-pub struct ProxyTimedOut(());
+    fn layer(&self, inner: M) -> Self::Service {
+        MakeTimeout { inner }
+    }
+}
 
-impl<M> tower::layer::Layer<M> for Layer {
-    type Service = Stack<M>;
+impl<T, M> NewService<T> for MakeTimeout<M>
+where
+    M: NewService<T>,
+    T: HasTimeout,
+{
+    type Service = Timeout<M::Service>;
 
-    fn layer(&self, inner: M) -> Self::Service {
-        Stack { inner }
+    fn new_service(&self, target: T) -> Self::Service {
+        match target.timeout() {
+            Some(t) => Timeout::new(self.inner.new_service(target), t),
+            None => Timeout::passthru(self.inner.new_service(target)),
+        }
     }
 }
 
-impl<T, M> tower::Service<T> for Stack<M>
+impl<T, M> tower::Service<T> for MakeTimeout<M>
 where
     M: tower::Service<T>,
     T: HasTimeout,
 {
-    type Response = tower::util::Either<Service<M::Response>, M::Response>;
+    type Response = Timeout<M::Response>;
     type Error = M::Error;
     type Future = MakeFuture<M::Future>;
```

Comment:
To go along with the renamings taking place, maybe call hits `MakeTimeoutFuture` now?

Reply:
I think _MakeFuture_ is descriptive enough -- we typically use _ResponseFuture_ in other contexts... I may change my mind on this later ;)